RavenDB-20090 Compacting LOH in next GC background run if LOH takes too much memory

arekpalinski opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20090/Consider-compacting-LOH-in-next-GC-background-run-after-Low-Memory-mode

### Additional description

During the regular memory status checkup we'll also check the size of Large Object Heap (every 15 minutes). If it takes more than 25% (configurable) we'll set `GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce` that will force LOH compaction during next blocking gen 2 garbage collection. 

This way we'll be able to recover from LOH fragmentation issues which resulted in OOM in some specific usage scenarios. Note that we don't do it during low memory because our experiments showed that it was too late already.

### Type of change

- Managed memory usage optimization

### How risky is the change?

- Moderate 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- Yes. It might affect any feature since it's about process wide usage of managed memory

### UI work

- No UI work is needed

arekpalinski committed:
RavenDB-20090 Compact LOH its size exceeds 25% of total memory and we're in low memory

arekpalinski committed:
RavenDB-20090 Remove the requirement for LOH compaction only if we're in low memory. Exposide the size of LOH in SNMP.

arekpalinski committed:
RavenDB-20090 Introduced a dedicated configuration option for LOH compaction threshold (in order to disable it completely you can provide -1). Making sure we're not setting CompactOnce if it's already set.

grisha-kotler commented on src/Sparrow/LowMemory/LowMemoryNotification.cs:
```diff
@@ -46,19 +47,53 @@ internal class LowMemEventDetails
                                            $"{MemoryUtils.GetExtendedMemoryInfo(memoryInfo)}");
                     }
+#if NET7_0_OR_GREATER
+                    if (now - _lastLohCompactionRequest > _lohCompactionInterval && GCSettings.LargeObjectHeapCompactionMode != GCLargeObjectHeapCompactionMode.CompactOnce)
```
Can we move this to a separate method?

arekpalinski replied:
Done

arekpalinski committed:
RavenDB-20090 Moving code checking if we need to compact LOH to the separate method

ppekrol commented on src/Sparrow/LowMemory/LowMemoryNotification.cs:
```diff
@@ -111,6 +121,39 @@ private void RunLowMemoryHandlers(bool isLowMemory, MemoryInfoResult memoryInfo,
+#if NET7_0_OR_GREATER
+        private void RequestLohCompactionIfNeeded(MemoryInfoResult memoryInfo, DateTime now)
+        {
+            if (now - _lastLohCompactionRequest > _lohCompactionInterval && GCSettings.LargeObjectHeapCompactionMode != GCLargeObjectHeapCompactionMode.CompactOnce)
```
we probably do not want that in Client API?

arekpalinski replied:
How can I check that?

arekpalinski replied:
Done in https://github.com/ravendb/ravendb/pull/16456/commits/249356027e3c12c28b8808d80c9048af90722837

arekpalinski committed:
RavenDB-20090 LOH compaction is supported only by LowMemoryNotification on Raven.Server (so it won't run on client side)

arekpalinski committed:
RavenDB-20090 Few fixes:
- avoid passing `threshold` to float.TryParse() since it will set 0 value when returning false
- better formatting to avoid deep code nesting
- enabling LOH compaction when running tests
